src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,429 @@
+        } else {
+            header_pointers[mid + 1].1
+        };
+        let amount = (next_offset - cur_offset) / self.header.record_header_size as u64;
+        (cur_offset, amount)
+    }
+
+    async fn read_header(&self, offset: u64) -> Result<RecordHeader> {
+        let mut buf = vec![0u8; self.header.record_header_size];
+        self.file.read_at(&mut buf, offset).await?;
+        deserialize(&buf).map_err(Into::into)
+    }
+
+    async fn read_headers(&self, offset: u64, amount: usize) -> Result<Vec<RecordHeader>> {
+        let mut buf = vec![0u8; self.header.record_header_size * amount];
+        self.file.read_at(&mut buf, offset).await?;
+        buf.chunks(self.header.record_header_size).try_fold(
+            Vec::with_capacity(amount),
+            |mut acc, bytes| {
+                acc.push(deserialize(&bytes)?);
+                Ok(acc)
+            },
+        )
+    }
+
+    async fn validate_header(&self, buf: &mut Vec<u8>) -> Result<()> {
+        if self.header.written != 1 {
+            return Err(Error::validation("header was not written").into());
+        }
+        if !Self::hash_valid(&self.header, buf)? {
+            return Err(Error::validation("header hash mismatch").into());
+        }
+        if self.header.version != HEADER_VERSION {
+            return Err(Error::validation("header version mismatch").into());
+        }
+        Ok(())
+    }
+
+    fn hash_valid(header: &IndexHeader, buf: &mut Vec<u8>) -> Result<bool> {
+        let hash = header.hash.clone();
+        let mut header = header.clone();
+        header.hash = vec![0; ring::digest::SHA256.output_len];
+        header.written = 0;
+        serialize_into(buf.as_mut_slice(), &header)?;
+        let new_hash = get_hash(&buf);
+        Ok(hash == new_hash)
+    }
+
+    async fn read_index_header(file: &File) -> Result<IndexHeader> {
+        let header_size = IndexHeader::serialized_size_default()? as usize;
+        let mut buf = vec![0; header_size];
+        file.read_at(&mut buf, 0).await?;
+        IndexHeader::from_raw(&buf).map_err(Into::into)
+    }
+
+    async fn read_tree_meta(file: &File, header: &IndexHeader) -> Result<TreeMeta> {
+        let meta_size = TreeMeta::serialized_size_default()? as usize;
+        let mut buf = vec![0; meta_size];
+        let fsize = header.filter_buf_size as u64;
+        let hs = header.serialized_size()?;
+        let meta_offset = hs + fsize + (header.records_count * header.record_header_size) as u64;
+        file.read_at(&mut buf, meta_offset).await?;
+        TreeMeta::from_raw(&buf).map_err(Into::into)
+    }
+
+    fn serialize(
+        headers_btree: &InMemoryIndex,
+        filter: &Bloom,
+    ) -> Result<Option<(IndexHeader, TreeMeta, Vec<u8>)>> {
+        debug!("blob index bptree serialize headers");
+        if let Some(record_header) = headers_btree.values().next().and_then(|v| v.first()) {
+            let record_header_size = record_header.serialized_size().try_into()?;
```

Comment:
IMO it's better to use builder pattern here, with object holding data for all methods needed. You can even add simple enum of states to ensure proper methods order (like "add_nonleaves" can be used only after "add_leaves" etc)

Reply:
I implemented it, but need your review (does it look strange?)